src/CMakeLists.txt

```diff
@@ -10,12 +10,20 @@ set(ENABLE_SINC_FAST_CONVERTER ${LIBSAMPLERATE_ENABLE_SINC_FAST_CONVERTER} PAREN
 set(ENABLE_SINC_MEDIUM_CONVERTER ${LIBSAMPLERATE_ENABLE_SINC_MEDIUM_CONVERTER} PARENT_SCOPE)
 set(ENABLE_SINC_BEST_CONVERTER ${LIBSAMPLERATE_ENABLE_SINC_BEST_CONVERTER} PARENT_SCOPE)
 
+include(CheckCSourceCompiles)
 include(ClipMode)
 include(CMakePackageConfigHelpers)
 
 # This will set CPU_CLIPS_NEGATIVE and CPU_CLIPS_POSITIVE
 clip_mode()
 
+# check for symbol visibility attributes support
+set(OLD_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS}")
+set(CMAKE_REQUIRED_FLAGS "-fvisibility=hidden -Werror ${OLD_REQUIRED_FLAGS}")
+check_c_source_compiles("int foo(void) __attribute__((visibility(\"default\")));
+                         int main(void) {return 0;}" HAVE_VISIBILITY)
+set(CMAKE_REQUIRED_FLAGS "${OLD_REQUIRED_FLAGS}")
```

Comment:
I would use [CMakePushCheckState](https://cmake.org/cmake/help/latest/module/CMakePushCheckState.html).

Reply:
Per your suggestion I changed it to this:
```
cmake_push_check_state()
set(CMAKE_REQUIRED_FLAGS "-fvisibility=hidden -Werror")
check_c_source_compiles("int foo(void) __attribute__((visibility(\"default\")));
                         int main(void) {return 0;}" HAVE_VISIBILITY)
cmake_pop_check_state()
```

It seemed to work, _however_ when I used a toolchain file, like
`-DCMAKE_TOOLCHAIN_FILE=Toolchain-MinGW64.cmake` it failed:
```
-- Checking processor clipping capabilities...
-- Checking processor clipping capabilities... disabled
CMake Error at src/CMakeLists.txt:21 (cmake_push_check_state):
  Unknown CMake command "cmake_push_check_state".


-- Configuring incomplete, errors occurred!
```

This is with cmake 3.9.6.  If it matters, the toolchain file looks like this:
```
# toolchain file I use to cross compile on Linux
# targetting Windows (x64/mingw-w64). running:
# cmake -DCMAKE_TOOLCHAIN_FILE=/path/to/Toolchain-MinGW64.cmake ....

SET(CMAKE_SYSTEM_NAME Windows)

SET(CMAKE_C_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-gcc)
SET(CMAKE_CXX_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-g++)
SET(CMAKE_RC_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-windres)

# where is the target environment
SET(CMAKE_FIND_ROOT_PATH /opt/cross_win64 /opt/cross_win64/x86_64-w64-mingw32)

# adjust the default behaviour of the FIND_XXX() commands:
# search headers and libraries in the target environment, search
# programs in the host environment
SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
SET(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

Reply:
@sezero what's the gain over the simple "whitelist" approach we currently have? I'm not opposed, but we can cover 98% of users with the current macro cases and don't have to pull in a bunch of cmake and autoconf macros.

Reply:
> @sezero what's the gain over the simple "whitelist" approach we currently have? I'm not opposed, but we can cover 98% of users with the current macro cases and don't have to pull in a bunch of cmake and autoconf macros.

IMO, a whitelist/blacklist approach should be against a well-known well-tested common case provided by autoconf or cmake.

Reply:
Why not see how the CMake authors did it: https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/GenerateExportHeader.cmake

`_test_compiler_hidden_visibility` macro looks interesting. In fact, I strongly suspect that we might be using a `COMPILER_HAS_HIDDEN_VISIBILITY` variable.

Reply:
I found what I did wrong with CMakePushCheckState: I had missed a cmake include, now it works.

As for how the CMake authors did it: I don't see how my three lines are inferior or wrong?

Force-pushed my last version.

Reply:
There are some corner cases, but they are rare. We can use current version.

Reply:
> There are some corner cases, but they are rare. We can use current version.

OK, anything else needed from me?